File: packages/rslint-test-tools/tests/typescript-eslint/rules/no-useless-default-assignment.test.ts
import { RuleTester } from '@typescript-eslint/rule-tester';

import { getFixturesRootDir } from '../RuleTester';

const rootDir = getFixturesRootDir();
const ruleTester = new RuleTester({
  languageOptions: {
    parserOptions: {
      project: './tsconfig.json',
      tsconfigRootDir: rootDir,
    },
  },
});

ruleTester.run('no-useless-default-assignment', {
  valid: [
    `
function withOptional(a: number | undefined = 1) {
  return a;
}
    `,
    `
const { a = 1 }: { a?: number } = {};
    `,
    `
[1, 2, 3, undefined].map((a = 42) => a + 1);
    `,
    `
function test(a: any = 'default') {
  return a;
}
    `,
  ],
  invalid: [
    {
      code: `
function basic(a: number = 1) {
  return a;
}
      `,
      errors: [{ messageId: 'uselessDefaultAssignment' }],
    },
    {
      code: `
function optional(a: number | undefined = undefined) {
  return a;
}
      `,
      errors: [{ messageId: 'preferOptionalSyntax' }],
    },
    {
      code: `
const { a = undefined }: { a?: number } = {};
      `,
      errors: [{ messageId: 'uselessUndefined' }],
    },
    {
      code: `
function withObject({ foo = '' }: { foo: string }) {
  return foo;
}
      `,
      errors: [{ messageId: 'uselessDefaultAssignment' }],
    },
    {
      code: `
[1, 2, 3].map((a = 42) => a + 1);
      `,
      errors: [{ messageId: 'uselessDefaultAssignment' }],
    },
    {
      code: `
interface B {
  foo: (b: boolean | string) => void;
}

const h: B = {
  foo: (b = false) => {},
};
      `,
      errors: [{ messageId: 'uselessDefaultAssignment' }],
    },
    {
      code: `
const [a = undefined] = [];
      `,
      errors: [{ messageId: 'uselessUndefined' }],
    },
  ],
});

File: internal/plugins/typescript/rules/no_useless_default_assignment/no_useless_default_assignment_test.go
package no_useless_default_assignment

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/typescript/rules/fixtures"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoUselessDefaultAssignmentRule(t *testing.T) {
	rule_tester.RunRuleTester(fixtures.GetRootDir(), "tsconfig.json", t, &NoUselessDefaultAssignmentRule, []rule_tester.ValidTestCase{
		{Code: `function withOptional(a: number | undefined = 1) { return a; }`},
		{Code: `const { a = 1 }: { a?: number } = {};`},
		{
			Code:     `const value = 1;`,
			TSConfig: "tsconfig.unstrict.json",
			Options: map[string]any{
				"allowRuleToRunWithoutStrictNullChecksIKnowWhatIAmDoing": true,
			},
		},
	}, []rule_tester.InvalidTestCase{
		{
			Code:   `function basic(a: number = 1) { return a; }`,
			Output: []string{`function basic(a: number) { return a; }`},
			Errors: []rule_tester.InvalidTestCaseError{{MessageId: "uselessDefaultAssignment"}},
		},
		{
			Code:   `function optional(a: number | undefined = undefined) { return a; }`,
			Output: []string{`function optional(a?: number | undefined) { return a; }`},
			Errors: []rule_tester.InvalidTestCaseError{{MessageId: "preferOptionalSyntax"}},
		},
		{
			Code:   `const { a = undefined }: { a?: number } = {};`,
			Output: []string{`const { a }: { a?: number } = {};`},
			Errors: []rule_tester.InvalidTestCaseError{{MessageId: "uselessUndefined"}},
		},
		{
			Code:   `function withObject({ foo = "" }: { foo: string }) { return foo; }`,
			Output: []string{`function withObject({ foo }: { foo: string }) { return foo; }`},
			Errors: []rule_tester.InvalidTestCaseError{{MessageId: "uselessDefaultAssignment"}},
		},
		{
			Code: `
interface B {
	foo: (b: boolean | string) => void;
}

const h: B = {
	foo: (b = false) => {},
};
			`,
			Output: []string{`
interface B {
	foo: (b: boolean | string) => void;
}

const h: B = {
	foo: (b) => {},
};
			`},
			Errors: []rule_tester.InvalidTestCaseError{{MessageId: "uselessDefaultAssignment"}},
		},
		{
			Code:     `const value = 1;`,
			TSConfig: "tsconfig.unstrict.json",
			Errors:   []rule_tester.InvalidTestCaseError{{MessageId: "noStrictNullCheck"}},
		},
	})
}
